### PR TITLE
Diarizerfast

### PR DESCRIPTION
**Pull request by Alex-Wengg**

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -13,14 +13,26 @@ FluidAudio is a speaker diarization system for Apple platforms using Core ML mod
...
 
-### DO NOT USE EMOJIs
-
-- Avoid using emojis as much as possible
-
-### Avoid redundant comments
-
-- Do not leave comments in places that are redundant or its very obvious
-- If a PR review asks a question, it may be a good indicator that the PR needs to add a comment there
+### ⚠️ NEVER CREATE DUMMY MODELS OR SYNTHETIC DATA
```
Comment (Alex-Wengg) [inline CLAUDE.md:16]: during my r&d, the model would love to bs "results" whenever it struggles so it can spend hours on something only for it be a big waste of time

File: CLAUDE.md [hunk 2]
```diff
@@ -13,14 +13,26 @@ FluidAudio is a speaker diarization system for Apple platforms using Core ML mod
...
+
+- **DO NOT** create dummy models for testing or development
+- **DO NOT** generate synthetic audio data for testing
+- **DO NOT** use random/fake models as placeholders
+- **DO NOT** create "demonstration" or "simulated" models that don't contain real weights
+- Always use the actual pyannote/wespeaker models
```
Comment (Alex-Wengg) [inline CLAUDE.md]: do you have suggestions for how we could generalize this ? sometimes it would make a "Mock" model

File: Sources/FluidAudioCLI/Benchmarks/BenchmarkRunner.swift [hunk 3]
```diff
@@ -119,10 +119,10 @@ struct BenchmarkRunner {
                 totalJER += metrics.jer
                 processedFiles += 1
```
Comment (Alex-Wengg) [inline Sources/FluidAudioCLI/Benchmarks/BenchmarkRunner.swift:120]: RTFx is much more clearer

**Comment by BrandonWeng** [comment]
> **RTF**	**22.62x**	<1.0x	⚠️	Real-Time Factor (lower is faster)

You probably need to update the if statement on the right too

File: CLAUDE.md (hunk 2 above)
Comment (BrandonWeng) [reply CLAUDE.md]: Just say to use the actual models in the code, don't make up random models? Don't have to be specifically pyannote/wespeaker models

**Comment by BrandonWeng** [comment]
@claude review

File: Sources/FluidAudio/Diarizer/ANEMemoryOptimizer.swift [hunk 4]
```diff
@@ -0,0 +1,264 @@
+import Accelerate
+import CoreML
+import Foundation
+import Metal
+
+/// ANE-optimized memory management for speaker diarization pipeline
+@available(macOS 13.0, iOS 16.0, *)
+public class ANEMemoryOptimizer {
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/ANEMemoryOptimizer.swift]: The code here seems mostly duplicated with the ones in the ASR model, is there a way to reduce the duplicate code? 

Perhaps a parent and extensions for each model type if needed?

Comment (BrandonWeng) [reply Sources/FluidAudio/Diarizer/ANEMemoryOptimizer.swift]: @claude compare this with the FluidAudio/ASR/ANEOptimizer.swift file

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 5]
```diff
@@ -30,6 +31,16 @@ public final class DiarizerManager {
...
+
+        // Initialize EmbeddingExtractor with wespeaker_v2 model
+        let cacheDir = DiarizerModels.defaultModelsDirectory()
+        let wespeakerPath = cacheDir.appendingPathComponent("wespeaker_v2.mlmodelc")
+        do {
+            self.embeddingExtractor = try EmbeddingExtractor(wespeakerPath: wespeakerPath)
+            logger.info("EmbeddingExtractor initialized with WeSpeaker V2 model")
+        } catch {
+            logger.error("Failed to initialize EmbeddingExtractor: \(error)")
+        }
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift]: Why is this here? We should have this in diarized models to be consistent. We could remove the non _v2 model from code but need to keep it on hugging face

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 6]
```diff
@@ -86,8 +96,12 @@ public final class DiarizerManager {
         let chunkSize = sampleRate * 10
         var allSegments: [TimedSpeakerSegment] = []
         var speakerDB: [String: [Float]] = [:]
+        var chunkIndex = 0
 
         for chunkStart in stride(from: 0, to: samples.count, by: chunkSize) {
+            chunkIndex += 1
+            logger.info("Processing chunk \(chunkIndex) at offset \(chunkStart)")
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift]:
```suggestion
```

Too verbose

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 7]
```diff
@@ -153,27 +162,54 @@ public final class DiarizerManager {
...
+                let speakerSum = slidingFeature.data[0][f].reduce(0, +)
+                let isClean: Float = speakerSum < 2.0 ? 1.0 : 0.0
+                speakerMask.append(slidingFeature.data[0][f][s] * isClean)
+            }
+            masks.append(speakerMask)
+        }
+
+        let embeddings = try embeddingExtractor.getEmbeddings(
+            audio: Array(paddedChunk),
+            masks: masks
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift]: @claude  Could you give a TLDR on how this works and why it's faster than the previous implementation?

File: Sources/FluidAudio/Diarizer/DiarizerModels.swift [hunk 8]
```diff
@@ -26,6 +26,25 @@ public struct DiarizerModels: Sendable {
...
+
+    var displayName: String {
+        return "WeSpeaker V2"
+    }
+
+    var isPackage: Bool {
+        fileName.hasSuffix(".mlpackage")
+    }
+}
+
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerModels.swift]: Make it consistent with the other models, we can just deprecate the older model from code entirely. No need to support the older vairent. Just update the 

`    private static let EmbeddingModelFileName = "wespeaker"`

File: Sources/FluidAudio/Diarizer/SegmentationProcessor.swift [hunk 9]
```diff
@@ -1,55 +1,177 @@
...
 
     private let logger = Logger(subsystem: "com.fluidinfluence.diarizer", category: "Segmentation")
+    private let memoryOptimizer = ANEMemoryOptimizer.shared
 
     public init() {}
 
     func getSegments(
-        audioChunk: ArraySlice<Float>, segmentationModel: MLModel, chunkSize: Int = 160_000
-    ) throws -> [[[Float]]] {
+        audioChunk: ArraySlice<Float>,
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/SegmentationProcessor.swift:16]: Might be best to add some unit tests for these changes too, including the clustering changes

Comment (BrandonWeng) [reply Sources/FluidAudio/Diarizer/SegmentationProcessor.swift:16]: Try to capture edge cases in regards to memory usage, they're quite fragile to changes

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 10]
```diff
@@ -153,27 +157,55 @@ public final class DiarizerManager {
...
-            binarizedSegments: binarizedSegments,
-            slidingWindowFeature: slidingFeature,
-            embeddingModel: models.embeddingModel,
-            sampleRate: sampleRate
+        // Use EmbeddingExtractor for embedding extraction
+        guard let embeddingExtractor = self.embeddingExtractor else {
+            throw DiarizerError.notInitialized
+        }
+
+        logger.info("Using EmbeddingExtractor for embedding extraction")
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift]:
```suggestion
        logger.deug("Using EmbeddingExtractor for embedding extraction")
```

File: Sources/FluidAudio/Diarizer/DiarizerModels.swift [hunk 11]
```diff
@@ -46,9 +48,10 @@ extension DiarizerModels {
         let directory = directory ?? defaultModelsDirectory()
         let config = configuration ?? defaultConfiguration()
 
+        // Download required models
         let modelNames = [
             SegmentationModelFileName + ".mlmodelc",
-            EmbeddingModelFileName + ".mlmodelc",
+            EmbeddingModelFileName + ".mlmodelc",  // wespeaker_v2.mlmodelc
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerModels.swift]:
```suggestion
            EmbeddingModelFileName + ".mlmodelc",  
```

File: Sources/FluidAudio/Diarizer/DiarizerModels.swift [hunk 12]
```diff
@@ -58,21 +61,28 @@ extension DiarizerModels {
             computeUnits: config.computeUnits
         )
 
-        guard let segmentationModel = models[SegmentationModelFileName + ".mlmodelc"],
-            let embeddingModel = models[EmbeddingModelFileName + ".mlmodelc"]
-        else {
+        // Load segmentation model
+        guard let segmentationModel = models[SegmentationModelFileName + ".mlmodelc"] else {
+            throw DiarizerError.modelDownloadFailed
+        }
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerModels.swift]: nit, you already defined the model + modelc in line 51-54?

File: Sources/FluidAudio/Diarizer/DiarizerModels.swift [hunk 13]
```diff
@@ -58,21 +61,28 @@ extension DiarizerModels {
...
         let endTime = Date()
         let totalDuration = endTime.timeIntervalSince(startTime)
-        // For now, we don't have separate download vs compilation times, so we'll estimate
-        // In reality, if models are cached, download time is 0
         let downloadDuration: TimeInterval = 0  // Models are typically cached
-        let compilationDuration = totalDuration  // Most time is spent on compilation
+        let compilationDuration = totalDuration
+
+        // Log model loading status
+        logger.info("Model Loading Status - Embedding Model: WeSpeaker V2")
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerModels.swift]:
```suggestion
```

File: Sources/FluidAudio/Diarizer/DiarizerTypes.swift [hunk 14]
```diff
@@ -1,12 +1,20 @@
 import Foundation
 
+public enum EmbeddingMode: String, Sendable, CaseIterable {
+    case full = "full"  // Full embedding extraction (10.5s)
+    case fast = "fast"  // Fast mode with optimizations
+    case skip = "skip"  // Skip embedding extraction entirely
+    case cached = "cached"  // Use cached embeddings if available
+}
```
Comment (BrandonWeng) [inline Sources/FluidAudio/Diarizer/DiarizerTypes.swift]: Do we need these types in this PR? This is probably better suited for the streaming change?

File: Tests/FluidAudioTests/StreamingAsrManagerTests.swift [hunk 15]
```diff
@@ -122,92 +122,87 @@ final class StreamingAsrManagerTests: XCTestCase {
...
-
-        let buffer = AVAudioPCMBuffer(
-            pcmFormat: format,
-            frameCapacity: AVAudioFrameCount(sampleRate * duration)
-        )!
-        buffer.frameLength = buffer.frameCapacity
-
-        // Stream audio (should not throw)
-        await manager.streamAudio(buffer)
+        // let manager = StreamingAsrManager()
```
Comment (BrandonWeng) [inline Tests/FluidAudioTests/StreamingAsrManagerTests.swift]: ?

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 16]
```diff
@@ -86,8 +92,11 @@ public final class DiarizerManager {
         let chunkSize = sampleRate * 10
         var allSegments: [TimedSpeakerSegment] = []
         var speakerDB: [String: [Float]] = [:]
+        var chunkIndex = 0
```
Comment (Bharat0091) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift]: not sure you are using it.

File: Sources/FluidAudio/Diarizer/DiarizerManager.swift [hunk 17]
```diff
@@ -153,27 +157,55 @@ public final class DiarizerManager {
...
+
+        let segmentationTime = Date().timeIntervalSince(segmentationStartTime)
+
+        // Unified and merged models removed - caused performance/stability issues
+
+        // Otherwise use traditional separate model processing
         let slidingFeature = segmentationProcessor.createSlidingWindowFeature(
             binarizedSegments: binarizedSegments, chunkOffset: chunkOffset)
 
-        let segmentationTime = Date().timeIntervalSince(segmentationStartTime)
```
Comment (Bharat0091) [inline Sources/FluidAudio/Diarizer/DiarizerManager.swift:163]: should segmentationTime include createSlidingWindowFeature ? [Though I think it should be taking very minimal time]

**Review by BrandonWeng — APPROVED** [review]
lgtm
